### PR TITLE
Enabled <<namePrefix>> token across modules and dependency pipeline resources

### LINKED ISSUE #909: Token Replacement: Using naming prefixes to support forking and instantiating the repo for CARML consumers
## Description

Currently a lot of the resources in the CARML repo use static names for resources (i.e. dependency pipeline resources and resources that must have unique name across Azure like Storage and key vault). We need to brainstorm how we can leverage the existing token replacement service to create name prefixes (i.e. using local tokens in the settings file or another method). And then implement it.

### PR DESCRIPTION
**Pull request by ahmadabdalla**
Closes #909

# Change

- Updated Modules & Dependency pipeline parameter files to use the new `<<namePrefix>>` token.
- Added token replacement for PowerShell tasks for dependency pipeline for both GH and ADO pipelines.
- Added new pester test to detect if parameter files contain the value for the `namePrefix` instead of the token.
- Added section in the getting started documentation to guide users in replacing the `namePrefix` token as one of the first PRs upon forking.

## Testing

- Tested ADO dependency pipeline after updating tasks. See [result](https://dev.azure.com/servicescode/infra-as-code-source/_build/results?buildId=164509&view=results)
- Tested GH dependency pipeline after updating tasks. [.Platform: Dependencies](https://github.com/Azure/ResourceModules/actions/workflows/platform.dependencies.yml)

- Tested different Modules to validate tokens functionality (Storage Account, Container Registries, Service Bus Namespace)

## Type of Change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: arm/.global/global.module.tests.ps1 [hunk 1]
```diff
@@ -763,6 +763,19 @@ Describe 'Deployment template tests' -Tag Template {
                 $TenantIdKeyCount -eq $TenantIdValueCount | Should -Be $true -Because ("Parameter file should not contain the Tenant ID guid, instead should reference a token value '$ParameterFileTokenName'")
             }
         }
+
+        It '[<moduleFolderName>] [Tokens] Parameter files, resource name should contain the namePrefix token' -TestCases $deploymentFolderTestCases {
```
Comment (MrMCake) [inline arm/.global/global.module.tests.ps1:767]:
```suggestion
        It '[<moduleFolderName>] [Tokens] Parameter files should use the namePrefix token where possible' -TestCases $deploymentFolderTestCases {
```

File: arm/.global/global.module.tests.ps1 [hunk 2]
```diff
@@ -763,6 +763,19 @@ Describe 'Deployment template tests' -Tag Template {
...
+            param (
+                [hashtable[]] $ParameterFileTestCases
+            )
+            foreach ($ParameterFileTestCase in $ParameterFileTestCases) {
+                $ParameterFileTokenName = -join ($ParameterFileTestCase.tokenSettings.tokenPrefix, 'namePrefix', $ParameterFileTestCase.tokenSettings.tokenSuffix)
+                $ParameterFileContent = Get-Content -Path $ParameterFileTestCase.parameterFile_Path
+                $namePrefixValue = $ParameterFileTestCase.tokenSettings.localTokens.tokens | Where-Object -Property name -EQ 'namePrefix' | Select-Object -ExpandProperty value
+                $namePrefixValueCount = ($ParameterFileContent | Select-String -Pattern "$namePrefixValue" -AllMatches).Matches.Count
+                $namePrefixValueCount -eq 0 | Should -Be $true -Because ("Parameter file should contain the token: $ParameterFileTokenName instead of value: $namePrefixValue")
```
Comment (MrMCake) [inline arm/.global/global.module.tests.ps1:776]: @ahmadabdalla , how big of a chance you see in this 'accidently' breaking deployments? Just thougt of a prefix like `kvlt` (dumb example, I know) - could that not potentially clash with just a resource type reference in a parameter name (e.g. if somebody uses `<namePrefix>-az-kvlt-min>`?

File: docs/wiki/GettingStarted.md [hunk 3]
```diff
@@ -141,6 +142,10 @@ As the modules we test oftentimes have dependencies to other services, we create
 
 Several fundamental variables are shared among all pipelines and are stored in a pipeline variable file. In case you want to not only leverage the module templates but actually re-use the implemented pipelines & testing framework as well, you need to set up several [variables](./PipelinesDesign#pipeline-variables) & [secrets](./PipelinesDesign#pipeline-secrets) in your environment.
 
+### Name Prefix Token Replacement
+
+Change the default `namePrefix` token value in the [Settings.json](https://github.com/Azure/ResourceModules/blob/main/settings.json) to a 3-5 character string that is used to distinguish your resources names. The default `namePrefix` token has a value `sxx`. Consider this to be one of the first pull requests you need to perform before running any of the repository pipelines. More information on tokens is provided [below](#parameter-file-tokens).
```
Comment (MrMCake) [inline docs/wiki/GettingStarted.md:147]: Can we confirm 5 characters is 'short' enough to not run into resource name limitations?

Comment (ahmadabdalla) [reply docs/wiki/GettingStarted.md:147]: Previously, i used `carml` as a test name in my test environment and did not run into issues back then. However reviewing [Resource name limits](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules) documentation. and cross referencing the ones that are risky with our repo, only one concerning is the virtual machine (windows), as it is 15 characters, maybe we can reduce the name [here](https://github.com/Azure/ResourceModules/blob/main/arm/Microsoft.Compute/virtualMachines/.parameters/windows.parameters.json) from 001 --> 1. 5 characters is a good threshold and gives wiggle room for people to make it more unique

File: arm/.global/global.module.tests.ps1 (hunk 2 above)
Comment (ahmadabdalla) [reply arm/.global/global.module.tests.ps1:776]: as discussed. Good point, I will remove the pester test for now, and open up an issue for discovery as a backlog item. https://github.com/Azure/ResourceModules/issues/951

File: arm/.global/global.module.tests.ps1 (hunk 1 above)
Comment (ahmadabdalla) [reply arm/.global/global.module.tests.ps1:767]: Will be removed from the tests and created as a backlog item

File: .github/workflows/platform.dependencies.yml [hunk 4]
```diff
@@ -386,17 +398,31 @@ jobs:
...
+            }
+
             # Retrieving parameters from previous job outputs and parameter files
             Write-Verbose "Retrieving parameters from previous job outputs" -Verbose
             $imageTemplateName = '${{ needs.job_deploy_imgt.outputs.imageTemplateName }}'
             $imageTemplateResourceGroup = '${{ needs.job_deploy_imgt.outputs.imageTemplateResourceGroup }}'
 
             Write-Verbose "Retrieving parameters from storage account parameter files" -Verbose
             $parameterFilePath = Join-Path $env:GITHUB_WORKSPACE '${{ env.dependencyPath }}' '${{ env.saNamespace }}' 'parameters' 'parameters.json'
+            $null = Convert-TokensInParameterFile @ConvertTokensInputs -ParameterFilePath $parameterFilePath -Verbose
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:420]: Should we run the conversion on all parameter files or twice on each parameter file as done for other tasks?

Comment (ahmadabdalla) [reply .github/workflows/platform.dependencies.yml:420]: as in creating a task dedicated to process all parameter files in the stage / job? I thought of another option is to construct the parameter files as an array, and then do a foreach. And if we need to do more files, we just add them to the array. To my knowledge, changes do not propagate between jobs, so we need an explicit task per job right?

Comment (eriqua) [reply .github/workflows/platform.dependencies.yml:420]: Nevermind I just see this is solved by lines 420 and 425. Resolving the comment

File: .github/workflows/platform.dependencies.yml [hunk 5]
```diff
@@ -386,17 +398,31 @@ jobs:
         uses: azure/powershell@v1
         with:
           inlineScript: |
+
+            # Load used functions
+            . (Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+            # Prepare Replace Tokens in Parameter File
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:405]:
```suggestion
            # Prepare replace tokens in parameter file
```

File: .github/workflows/platform.dependencies.yml [hunk 6]
```diff
@@ -933,8 +979,22 @@ jobs:
         uses: azure/powershell@v1
         with:
           inlineScript: |
-            # Get key vault name
+
             $parameterFilePath = Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'dependencies' '${{ env.namespace }}' 'parameters' 'sqlmi.parameters.json'
+            # Load used functions
+            . (Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+            # Replace Tokens in Parameter File
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:987]:
```suggestion
            # Replace tokens in parameter file
```

File: .github/workflows/platform.dependencies.yml [hunk 7]
```diff
@@ -243,11 +243,23 @@ jobs:
         uses: azure/powershell@v1
         with:
           inlineScript: |
+
+            $parameterFilePath = Join-Path $env:GITHUB_WORKSPACE '${{ env.dependencyPath }}' '${{ env.namespace }}' 'parameters' 'parameters.json'
             # Load used functions
             . (Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'sharedScripts' 'Export-ContentToBlob.ps1')
+            . (Join-Path $env:GITHUB_WORKSPACE 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+            # Replace Tokens in Parameter File
```
Comment (eriqua) [inline .github/workflows/platform.dependencies.yml:252]:
```suggestion
            # Replace tokens in parameter file
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 8]
```diff
@@ -186,11 +186,23 @@ stages:
               azureSubscription: $(serviceConnection)
               ScriptType: 'InlineScript'
               Inline: |
+
+                $parameterFilePath = Join-Path '$(Build.SourcesDirectory)' '$(dependencyPath)' '$(resourceType)' 'parameters' 'parameters.json'
                 # Load used functions
                 . (Join-Path '$(Build.SourcesDirectory)' 'utilities' 'pipelines' 'sharedScripts' 'Export-ContentToBlob.ps1')
+                . (Join-Path '$(Build.SourcesDirectory)' 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+                # Replace Tokens in Parameter File
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:195]:
```suggestion
                # Replace tokens in parameter file
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 9]
```diff
@@ -293,17 +305,31 @@ stages:
               azureSubscription: $(serviceConnection)
               ScriptType: 'InlineScript'
               Inline: |
+
+                # Load used functions
+                . (Join-Path '$(Build.SourcesDirectory)' 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+                # Prepare Replace Tokens in Parameter File
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:312]:
```suggestion
                # Prepare replace tokens in parameter file
```

File: .azuredevops/platformPipelines/platform.dependencies.yml [hunk 10]
```diff
@@ -661,8 +701,22 @@ stages:
                   azureSubscription: $(serviceConnection)
                   ScriptType: 'InlineScript'
                   Inline: |
-                    # Get key vault name
+
                     $parameterFilePath = Join-Path '$(Build.SourcesDirectory)' '$(dependencyPath)' '$(resourceType)' 'parameters' 'sqlmi.parameters.json'
+                    # Load used functions
+                    . (Join-Path '$(Build.SourcesDirectory)' 'utilities' 'pipelines' 'tokensReplacement' 'Convert-TokensInParameterFile.ps1')
+
+                    # Replace Tokens in Parameter File
```
Comment (eriqua) [inline .azuredevops/platformPipelines/platform.dependencies.yml:709]:
```suggestion
                    # Replace tokens in parameter file
```

File: docs/wiki/GettingStarted.md (hunk 3 above)
Comment (eriqua) [reply docs/wiki/GettingStarted.md:147]: I wouldn't reduce the vm number suffix that much. I'd either go with 4 chars or 001-->01 (15 chars should be fine and it's now 14)

**Review by MrMCake — APPROVED** [review]

**Pull request merged**